astropy/io/fits/tests/test_convenience.py

```diff
@@ -89,6 +91,68 @@ def test_table_to_hdu_convert_comment_convention(self):
         with pytest.raises(ValueError):
             hdu.header.index('comments')
 
+    def test_table_to_hdu_filter_reserved(self):
+        """
+        Regression test for https://github.com/astropy/astropy/issues/9387
+        """
+        table = Table([[1, 2, 3], ['a', 'b', 'c'], [2.3, 4.5, 6.7]],
+                      names=['a', 'b', 'c'], dtype=['i4', 'U1', 'f8'])
+        table.meta.update({'EXPTIME': 32.1, 'XTENSION': 'NEWTABLE',
+                           'NAXIS': 1, 'NAXIS1': 3, 'NAXIS2': 9,
+                           'PCOUNT': 42, 'OBSERVER': 'Adams'})
+        with pytest.warns(AstropyUserWarning, match=r'Meta-data keyword \w+ '
+                          r'will be ignored since it conflicts with a FITS '):
+            hdu = fits.table_to_hdu(table)
+
+        assert hdu.header.get('XTENSION') == 'BINTABLE'
+        assert hdu.header.get('NAXIS') == 2
+        assert hdu.header.get('NAXIS1') == 13
+        assert hdu.header.get('NAXIS2') == 3
+        assert hdu.header.get('PCOUNT') == 0
+        np.testing.assert_almost_equal(hdu.header.get('EXPTIME'), 3.21e1)
+
+    @pytest.mark.parametrize('card', REMOVE_KEYWORDS)
+    def test_table_to_hdu_warn_reserved(self, card):
+        """
+        Test warning for each keyword in ..connect.REMOVE_KEYWORDS, 1 by 1
+        """
+        diagnose = 'be ignored since it conflicts with a FITS reserved keyword'
+        res_cards = {'XTENSION': 'BINTABLE', 'BITPIX': 8,
+                     'NAXIS': 2, 'NAXIS1': 12, 'NAXIS2': 3,
+                     'PCOUNT': 0, 'GCOUNT': 1, 'TFIELDS': 2, 'THEAP': None}
+        ins_cards = {'XTENSION': 'TABLE', 'BITPIX': 16,
+                     'NAXIS': 1, 'NAXIS1': 2, 'NAXIS2': 6,
+                     'PCOUNT': 2, 'GCOUNT': 2, 'TFIELDS': 4, 'THEAP': 36}
+
+        table = Table([[1.0, 2.0, 3.0], [2.3, 4.5, 6.7]],
+                      names=['wavelength', 'flux'], dtype=['f8', 'f4'])
+        table.meta['ORIGIN'] = 'Min.Silly Walks'
+        table.meta[card] = ins_cards[card]
+        assert table.meta.get(card) != res_cards[card]
```

Comment:
I don't understand what you're doing here...
What I had in mind is that when doing `fits.table_to_hdu(table)` you can get several (possibly different) warnings, and `pytest.warns` does not allow to check which warning is raised.

Reply:
For instance:
```
In [2]:  
   ...:         table = Table([[1, 2, 3], ['a', 'b', 'c'], [2.3, 4.5, 6.7]], 
   ...:                       names=['a', 'b', 'c'], dtype=['i4', 'U1', 'f8']) 
   ...:         table.meta.update({'EXPTIME': 32.1, 'XTENSION': 'NEWTABLE', 
   ...:                            'NAXIS': 1, 'NAXIS1': 3, 'NAXIS2': 9, 
   ...:                            'PCOUNT': 42, 'OBSERVER': 'Adams'})          

In [3]: fits.table_to_hdu(table)                                                
WARNING: Meta-data keyword XTENSION will be ignored since it conflicts with a FITS reserved keyword [astropy.io.fits.convenience]
WARNING: Meta-data keyword NAXIS will be ignored since it conflicts with a FITS reserved keyword [astropy.io.fits.convenience]
WARNING: Meta-data keyword NAXIS1 will be ignored since it conflicts with a FITS reserved keyword [astropy.io.fits.convenience]
WARNING: Meta-data keyword NAXIS2 will be ignored since it conflicts with a FITS reserved keyword [astropy.io.fits.convenience]
WARNING: Meta-data keyword PCOUNT will be ignored since it conflicts with a FITS reserved keyword [astropy.io.fits.convenience]
```
Here I think it would be good to test that the warnings are the expected ones.

Reply:
Now I am not sure what kind of test you had in mind...
`pytest.warns(..., match=f'Meta-data keyword {card} will {diagnose}')`
does check for the specific keyword, just one at a time. Or did you intend to check them all at once, e.g. using `catch_warnings() as w` or `pytest.warns() as w` and then go through all the lines in `w`? Not sure if the warning lines will have a predetermined order, so this could be a bit trickier, and I do not see much advantage in this variant, except for checking that multiple reserved keywords indeed raise a warning for each of them.

Reply:
My point is to be sure that you get the expected number of warnings, 5 for the example above. Or said another way, that for each reserved keyword put in `table.meta` you get a warning.
So yes maybe with `catch_warnings() as w` (I don't know if this is possible with `pytest.warns` ?). And maybe using fewer reserved keywords in the test to simplify a bit ?

Reply:
Testing the number is simple; and as the warnings do appear to pop up in the order of the keys (which are always ordered in a 3.6+ dict, as @pllim has let me know ;-), it is even possible to test for the specific keywords. Not sure if we can rely on the warnings to stay in order, though.
Since the second test now checks each reserved keywords, I guess we can do here with fewer - my main objective was to have the `NAXIS` keywords covered.

Reply:
Oh, and this change assumes that no other warnings are emitted...

Reply:
I think that the order should stay, indeed. And I actually missed that `res|ins_cards` contain all the keywords in `REMOVE_KEYWORDS`. So I think we can go with this, thanks @dhomeier !

Reply:
Thanks for the review, @saimn !